python-frontend/src/test/java/org/sonar/python/tree/PythonTreeMakerTest.java

```diff
@@ -747,6 +749,54 @@ void nonlocalStatement() {
     assertThat(nonlocalStatement.children()).hasSize(3);
   }
 
+  @Test
+  void funcdef_statement_type_params() {
+    setRootRule(PythonGrammar.FUNCDEF);
+    var astNode = p.parse("def overly_generic[\n" +
+      "   SimpleTypeVar,\n" +
+      "   TypeVarWithBound: int,\n" +
+      "   TypeVarWithConstraints: (str, bytes),\n" +
+      "   *SimpleTypeVarTuple,\n" +
+      "   **SimpleParamSpec\n" +
+      "](\n" +
+      "   a: SimpleTypeVar,\n" +
+      "   b: TypeVarWithBound,\n" +
+      "   c: Callable[SimpleParamSpec, TypeVarWithConstraints],\n" +
+      "   *d: SimpleTypeVarTuple,\n" +
+      "): pass");
+    var functionDef = treeMaker.funcDefStatement(astNode);
+    assertThat(functionDef.name()).isNotNull();
+    var typeParams = functionDef.typeParams();
+    assertThat(typeParams).isNotNull();
+    assertThat(typeParams.getKind()).isEqualTo(Kind.TYPE_PARAMS);
+    assertThat(typeParams.leftBracket()).isNotNull();
+    assertThat(typeParams.rightBracket()).isNotNull();
+    assertThat(typeParams.children()).hasSize(11);
+    var typeParamsList = typeParams.typeParams();
+    assertThat(typeParamsList).isNotNull().hasSize(5);
+
+    var simpleTypeVar = typeParamsList.get(0);
+    assertThat(simpleTypeVar.name().name()).isEqualTo("SimpleTypeVar");
+    assertThat(simpleTypeVar.starToken()).isNull();
+    assertThat(simpleTypeVar.typeAnnotation()).isNull();
+
+    var typeWithBound = typeParamsList.get(1);
+    assertThat(typeWithBound.name().name()).isEqualTo("TypeVarWithBound");
+    assertThat(typeWithBound.starToken()).isNull();
+    assertThat(typeWithBound.typeAnnotation()).isNotNull();
```

Comment:
_Nitpick:_ Maybe we can have a small assertion to ensure the type annotation is correctly created.